tests/unit/models/watch-detector-test.js

```diff
@@ -0,0 +1,273 @@
+        };
+        // we assuming polling can never not work, if it doesn't sorry..
+        return subject.findBestWatcherOption({ watcher: 'node' }).then(function(option) {
+          expect(option).to.have.property('watcher', 'polling');
+          expect(ui.output).to.eql('was unable to use: "node", fell back to: "polling"' + EOL);
+        });
+      });
+
+      it('falls back to polling if unwatch fails', function() {
+        fs.watch = function() {
+          // works
+        };
+        fs.unwatch = function() {
+          throw new Error('OMG');
+        };
+        // we assuming polling can never not work, if it doesn't sorry..
+        return subject.findBestWatcherOption({ watcher: 'node' }).then(function(option) {
+          expect(option).to.have.property('watcher', 'polling');
+          expect(ui.output).to.eql('was unable to use: "node", fell back to: "polling"' + EOL);
+        });
+      })
+    });
+
+  });
+
+  describe('#checkWatchman', function() {
+    describe('watchmanSupportsPlatform', function() {
+      it('true: hides the "watchman not found, falling back to XYZ message"', function() {
+        subject.watchmanSupportsPlatform = true;
+
+        childProcess.exec = function() {
+          return Promise.reject();
+        };
+        fs.watch = function() {};
+        fs.unwatch = function() {};
+
+        return subject.checkWatchman().then(function(result) {
+          expect(result).to.have.property('watcher', 'node');
+          expect(ui.output).to.eql('');
+        });
+      });
+
+      it('false: shows the "watchman not found, falling back to XYZ message"', function() {
+        subject.watchmanSupportsPlatform = false;
+
+        childProcess.exec = function() {
+          return Promise.reject();
+        };
+        fs.watch = function() {};
+        fs.unwatch = function() {};
+
+        return subject.checkWatchman().then(function(result) {
+          expect(result).to.have.property('watcher', 'node');
+          expect(ui.output).to.match(/Could not start watchman/);
+          expect(ui.output).to.match(/Visit http:\/\/ember-cli.com\/user-guide\/\#watchman/);
+        });
+      });
+    });
+    it('prefers watchman if everything appears to be good', function() {
+      childProcess.exec = function() {
+        return Promise.resolve('{"version":"3.0.0"}');
+      };
+
+      return subject.checkWatchman().then(function(preference) {
+        expect(preference).to.have.property('watcher', 'watchman');
+        expect(ui.output).to.not.match(/Could not start watchman/);
+        expect(ui.output).to.not.match(/falling back to NodeWatcher/);
+        expect(ui.output).to.not.match(/ember-cli\.com\/user-guide\/\#watchman/);
+        expect(ui.output).to.not.match(/Looks like you have a different program called watchman/);
+        expect(ui.output).to.not.match(/Invalid watchman found/);
+      });
+    });
+
+    describe('fallse back to NODE', function() {
+      var iff = it;
```

Comment:
heh, they only read better if you know that `iif` means "if and only if". You CS people. 😈